src/sage/schemes/elliptic_curves/hom_sum.py

```diff
@@ -0,0 +1,644 @@
+            if self.is_endomorphism():
+                R,S = P,Q
+            else:
+                R,S = self.codomain().change_ring(F).torsion_basis(l**m)
+            M = self.matrix_on_subgroup((P,Q), (R,S))
+            g = ZZ(gcd(M.list())).p_primary_part(l)
+            if g > 1:
+                scalar *= g
+                M = (M.change_ring(ZZ) / g).change_ring(M.base_ring())
+            K = M.left_kernel_matrix()
+            for row in K:
+                u,v = map(ZZ, row)
+                pt = u*P + v*Q
+                pt.set_order(row.additive_order())
+                ker.append(pt)
+
+        from sage.schemes.elliptic_curves.hom_composite import EllipticCurveHom_composite
+        phi = EllipticCurveHom_composite(self.domain(), [])
+
+        if scalar != 1:
+            phi *= phi.codomain().scalar_multiplication(scalar)
+
+        while ker:
+            K = ker.pop(0)
+
+            (l,e), = K.order().factor()
+            for i in reversed(range(e)):
+                Kl = l**i * K
+                Kl.set_order(l)
+
+                from sage.groups.generic import multiples
+                from sage.misc.misc_c import prod
+                x = polygen(Kl.base_ring())
+                poly = prod(x - T.xy()[0] for T in multiples(Kl, l//2, Kl))
+                poly = poly.change_ring(self.base_ring())
+
+                psi = phi.codomain().isogeny(poly)
+                phi = psi * phi
+                K = psi._eval(K)
+                ker = [psi._eval(P) for P in ker]
+
+        if insep:
+            frob = phi.codomain().frobenius_isogeny(insep)
+            phi = frob * phi
+
+        from sage.schemes.elliptic_curves.hom import find_post_isomorphism
+        iso = find_post_isomorphism(phi, self)
+        return iso * phi
+
+    # EllipticCurveHom methods
+
+    def _degree_bound(self):
+        r"""
+        Return an upper bound on the degree of this sum morphism.
+
+        ALGORITHM: Induction using the bound
+        `|\deg(f+g) - \deg(f) - \deg(g)| \leq 2\sqrt{\deg(f)\cdot\deg(g)}`;
+        see for instance Lemma V.1.2 of [Sil2009]_.
+
+        EXAMPLES::
+
+            sage: E = EllipticCurve(GF(101), [5,5])
+            sage: phi = E.isogenies_prime_degree(7)[0]
+            sage: (phi + phi)._degree_bound()
+            28
+            sage: (phi + phi).degree()
+            28
+            sage: (phi - phi)._degree_bound()
+            28
+            sage: (phi - phi).degree()
+            0
+
+        ::
+
+            sage: E = EllipticCurve(GF(443), [1,1])
+            sage: pi = E.frobenius_endomorphism()
+            sage: m1 = E.scalar_multiplication(1)
+            sage: (pi - m1)._degree_bound()
+            486
+            sage: (pi - m1).degree()
+            433
+        """
+        B = ZZ(0)
+        for phi in self._phis:
+            B += phi.degree() + 2*sqrt(B * phi.degree())
+        return integer_floor(B)
+
+    def _compute_degree(self):
```

Comment:
I feel sure that there must be a way to compute the degree of the sum of two morphisms; if so you caould call it repeateldy here.  But I'll need to think about that a bit more.

Reply:
I suppose what definitely works is the identity $\deg(\varphi+\psi) = \deg(\varphi)+\deg(\psi)+\mathrm{tr}(\varphi\widehat\psi)$, now that we have #35546. But there should be cases where computing multiple such trace pairings will certainly be slower than the current one-shot method, so this might require some thinking...